### DISCUSSION TITLE
Audio visualization support (without implementing Web Audio API)

### DISCUSSION THREAD

**Discussion by TF3RDL** [TheQwertiest/foo_spider_monkey_panel#222] [Ideas]
A support for visualizations would be nice, and when combined with x64 support (#199), they can replace legacy visualization components that won't get updated for x64 versions of fb2k like Musical Spectrum (can be replaced with [this CodePen audio visualizer](https://codepen.io/TF3RDL/pen/poQJwRW) and [audioMotion-analyzer](https://github.com/hvianna/audioMotion-analyzer)) and Channel Spectrum panel (can be replaced with [this CodePen project of similar name](https://codepen.io/TF3RDL/pen/NWBYeaE)).

However, this has to be done through a non-standard API (similar to Wallpaper Engine implementing non-standard APIs for visualizations in web wallpapers) implementing visualization streams from fb2k SDK; the `visualization.getAbsoluteTime()` method can be used to retrieve time for visualization data and while `visualization.getChunkAbsolute(bufferSize, offset)` returns short chunk (with length of `bufferSize`) of waveform data at timestamp `offset` (when `offset` is not defined or omitted, the `getAbsoluteTime()` return value is used as default) represented as a stripped down version of `AudioBuffer` from Web Audio API that lacks playback capabilities and with the differences that `getChannelData()` method returns `Float64Array` waveform data and can accept strings as arguments representing audio channels.

BTW, I know implementing Web Audio API to SMP is pointless (except for visualizations), therefore I propose these nonstandard APIs above

**Comment by TheQwertiest** [comment]
Added a corresponding card in the project: https://github.com/users/TheQwertiest/projects/1/views/1?pane=issue&itemId=44126294
But, tbh, I don't think it will be implemented any time soon, since there are a lot of features with much higher priority (e.g. fetch API, built-in DPI scaling, Direct2D)
